### PR TITLE
Manual page generation

### LINKED ISSUE #848: cabal-install needs a manpage
(Imported from [Trac #858](http://hackage.haskell.org/trac/hackage/ticket/858), reported by @nomeata on 2011-06-19)

The cabal binary shipped by cabal-install ought to be documented with a manage. Ideally, cabal could generate one by itself from the Command declarations, similar to how darcs does it, this way there is no duplication of documentation.

### PR DESCRIPTION
**Pull request by mmakowski**
Manual page generation from command list defined in `Main`. Based on the approach and implementation in darcs.

Fixes #848.

### CODE REVIEW

**Comment by 23Skidoo** [comment]
Thanks! Looks good on a first glance.

File: cabal-install/Distribution/Client/Manpage.hs [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+  , manpage
+  ) where
+
+import Distribution.Simple.Command
+import Distribution.Client.Setup (globalCommand)
+
+import Data.Char (toUpper)
+import Data.List (intercalate)
+
+data CommandVisibility = Visible | Hidden
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/Manpage.hs]: Looks like this duplicates the functionality of `CommandType` in `D.S.Command`.

File: cabal-install/Distribution/Client/Manpage.hs [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+-- | wraps a @CommandUI@ together with a function that turns it into a @Command@.
+-- By hiding the type of flags for the UI allows construction of a list of all UIs at the
+-- top level of the program. That list can then be used for generation of manual page
+-- as well as for executing the selected command.
+data CommandSpec action
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/Manpage.hs]: This looks like something that belongs better in `D.S.Command`. Consider refactoring that module.

File: cabal-install/Distribution/Client/Manpage.hs [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+  = forall flags. CommandSpec (CommandUI flags) (CommandUI flags -> Command action) CommandVisibility
+
+commandFromSpec :: CommandSpec a -> Command a
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/Manpage.hs]: Ditto.

**Comment by 23Skidoo** [comment]
LGTM but I'd like to see some parts moved to `D.S.Command`.

**Comment by 23Skidoo** [comment]
A potential future improvement may be to add support for generating per-command manpages (e.g. `cabal manpage install`). Then we could make `cabal help` display a manpage when run from the terminal, a-la `git help`.

**Comment by 23Skidoo** [comment]
FYI, there are also some formatting problems (but I don't see them as impediment for merging):

formatting_problem

**Comment by mmakowski** [comment]
@23Skidoo, thanks for the review. I will try to address the comments over the next couple of days.

**Comment by 23Skidoo** [comment]
> Regarding `cabal manpage install`, I think that woud require making `manpage` a library-level special command similar to `help`.

I am not so sure. If you can generate a manpage with information about all commands from `cabal-install`, it should also be possible to do the same for a single command (maybe after some refactoring of `D.S.Command`).

Code for help formatting and option handling lives in the Cabal lib because we want to support things like `./Setup foo --help`.

> Another thing to consider with help displaying manual pages is that man is not availabe on Windows.

In these cases we can just revert to the old mode of output.

**Comment by mmakowski** [comment]
I have now moved command-related bits to `D.S.Command` and fixed formating of `sandbox` sub-commands.

**Comment by 23Skidoo** [comment]
LGTM now.

**Comment by 23Skidoo** [comment]
Merged, thanks!

**Pull request merged**